pkg/storegateway/indexheader/reader_pool.go

```diff
@@ -51,21 +59,64 @@ type ReaderPool struct {
 	lazyReaders   map[*LazyBinaryReader]struct{}
 }
 
-// NewReaderPool makes a new ReaderPool and starts a background task for unloading idle Readers if enabled.
-func NewReaderPool(logger log.Logger, lazyReaderEnabled bool, lazyReaderIdleTimeout time.Duration, metrics *ReaderPoolMetrics) *ReaderPool {
+// LazyLoadedHeadersSnapshotConfig stores information needed to track lazy loaded index headers.
+type LazyLoadedHeadersSnapshotConfig struct {
+	Path   string
+	UserID string
+}
+
+type lazyLoadedHeadersSnapshot struct {
+	// IndexHeaderLastUsedTime is map of index header ulid.ULID to timestamp in millisecond.
+	IndexHeaderLastUsedTime map[ulid.ULID]int64 `json:"index_header_last_used_time"`
+	UserID                  string              `json:"user_id"`
+}
+
+// persist atomically writes this snapshot to persistDir.
+func (l lazyLoadedHeadersSnapshot) persist(persistDir string) error {
+	// Create temporary path for fsync.
+	// We don't use temporary folder because the process might not have access to the temporary folder.
+	tmpPath := filepath.Join(persistDir, strings.Join([]string{"tmp", lazyLoadedHeadersListFile}, "-"))
```

Comment:
[nit] Couldn't we just do this?
```suggestion
	tmpPath := filepath.Join(persistDir, "tmp-"+lazyLoadedHeadersListFile)
```